### PR TITLE
Major performance improvment by bypassing the sequence of ports when sending tree data

### PR DESCRIPTION
**Pull request by clbond**
This PR represents a major performance improvement over previous builds.

Before this change, we were sending tree data through the normal channel. That meant every time we sent a tree or a tree delta, it had to go through this sequence:

Backend script -> content script -> background channel -> frontend

Each one of these steps required a serialize + deserialize (which happens in the internal Chrome code, not our code). Therefore for large trees, it was an absolute performance killer. This PR fixes that. Now the trees are just serialized and deserialized **once** on their way from the backend to the frontend.

I encourage anyone who reviews this to do a side-by-side performance comparison using the Kitchen Sink stress tester or any other app.

### CODE REVIEW

File: src/backend/backend.ts [hunk 1]
```diff
@@ -15,39 +14,67 @@ import {
...
 
 declare const ng;
 declare const getAllAngularRootElements: () => Element[];
 declare const treeRenderOptions: SimpleOptions;
 
+/// For tree deltas that contain more changes than {@link deltaThreshold},
+/// we simply send the entire tree again instead of trying to patch it
+/// since it will be faster than trying to apply hundreds or thousands of
+/// changes to an existing tree.
+const deltaThreshold = 512;
```
Comment (xorgy) [inline src/backend/backend.ts]: Did you profile to set this constant, or is it very ballpark?

Comment (clbond) [inline src/backend/backend.ts]: I don't have measurements. But I did do a bit of testing around it and it appeared to improve performance. Logically it also makes sense. There is no sense trying to compare two massive trees that you already know are vastly different. Just send the new one. Other than that does the PR look OK? Did you try running it?

**Comment by oskarols** [comment]
Tried testing it on a large RC4 app, component tree is showing up blank unfortunately. I'm assuming this only works for RC5 or 6?

**Comment by clbond** [comment]
@oskarols Are there any exceptions in the backend or the frontend? For backend, you can just open the Console tab that you would normally use for your app. For the frontend, open Augury inside a popup (not docked) and then hit Cmd-Shift-I (Mac) or Ctrl-Shift-I (Windows).

Is your app publicly available so I can reproduce?

**Comment by oskarols** [comment]
Sadly it's not publicly available. Seeing no exceptions in either.

**Comment by clbond** [comment]
If you can make a plunkr that reproduces the issue I would love to debug and fix it. I have used Augury for RC4 appals and it appeared to work properly. Alternatively if you have a bit of time to do a screen-sharing session with me, we could look into the issue together.

File: src/backend/connection.ts [hunk 2]
```diff
@@ -33,6 +33,24 @@ export const subscribe = (handler: MessageHandler) => {
...
+  switch (message.messageType) {
+    /// These types of messages should never be sent through this mechanism. A DispatchWrapper
+    /// message is for communication between content-script and the backend and has no business
+    /// being sent to the frontend. Similarly, a message containing tree data should be sent
+    /// through the {@link MessageBuffer} mechanism in backend.ts instead of through this port.
+    /// Sending a message with the {@link send} function will cause that message to take a very
+    /// circuitous route and will be serialized and deserialized repeatedly. Therefore large
+    /// messages must be sent using the {@link MessageBuffer} mechanism in order to avoid major
+    /// performance bottlenecks and UI latency.
+    case MessageType.CompleteTree:
```
Comment (igor-ka) [inline src/backend/connection.ts]: @clbond Do we need those 2 empty case conditions here?

Comment (clbond) [inline src/backend/connection.ts]: Yes, they fall through to the error handler. The alternative would be to do:

```
if (messageType == MessageType.TreeDiff || messageType == MessageType.CompleteTree || messageType == MessageType.DispatchWrapper) {
  throw ...;
}
```

File: src/backend/backend.ts (hunk 1 above)
Comment (igor-ka) [inline src/backend/backend.ts]: I am ok with 512 being a ballpark estimate at this point

File: src/backend/connection.ts (hunk 2 above)
Comment (clbond) [inline src/backend/connection.ts]: Fixed. Replaced with if statement

File: src/backend/backend.ts [hunk 3]
```diff
@@ -15,39 +14,67 @@ import {
...
+
+  /// Send a message through the normal channels to indicate to the frontend
+  /// that messages are waiting for it in {@link messageBuffer}
+  send<void, void>(MessageFactory.push());
 
-  send<void, any>(
-    previousTree
-      ? MessageFactory.treeDiff(previousTree.diff(newTree))
-      : MessageFactory.completeTree(newTree));
+  previousTree = tree;
```
Comment (igor-ka) [inline src/backend/backend.ts]: In the future let's extract the variable assignment out of it. One way is to have a redux store on the backend that will manage this. This is completely out of scope though.

Comment (clbond) [inline src/backend/backend.ts]: Alright, I will make a note of this to address in the near future.

**Comment by igor-ka** [comment]
@clbond LGTM :shipit:

**Comment by clbond** [comment]
Thank you Igor. I appreciate your thorough review :)

**Comment by clbond** [comment]
@oskarols Are you trying to debug an app that is running in production mode?

**Comment by oskarols** [comment]
@clbond Nope, not in production mode. Looked into it, and it seems we're getting some exception from Augury. 

```
zone.js:269 Uncaught TypeError: Cannot read property 'name' of undefinedexports.functionName @ function-name.ts?6954:7(anonymous function) @ transformer.ts?d0de:89dependencies @ transformer.ts?d0de:89(anonymous function) @ transformer.ts?d0de:137load @ transformer.ts?d0de:53exports.transform @ transformer.ts?d0de:61(anonymous function) @ transformer.ts?d0de:153transformChildren @ transformer.ts?d0de:151(anonymous function) @ transformer.ts?d0de:172load @ transformer.ts?d0de:53exports.transform @ transformer.ts?d0de:61(anonymous function) @ transformer.ts?d0de:153transformChildren @ transformer.ts?d0de:151(anonymous function) @ transformer.ts?d0de:172load @ transformer.ts?d0de:53exports.transform @ transformer.ts?d0de:61(anonymous function) @ transformer.ts?d0de:153transformChildren @ transformer.ts?d0de:151(anonymous function) @ transformer.ts?d0de:172load @ transformer.ts?d0de:53exports.transform @ transformer.ts?d0de:61(anonymous function) @ transformer.ts?d0de:153transformChildren @ transformer.ts?d0de:151(anonymous function) @ transformer.ts?d0de:172load @ transformer.ts?d0de:53exports.transform @ transformer.ts?d0de:61(anonymous function) @ transformer.ts?d0de:153transformChildren @ transformer.ts?d0de:151(anonymous function) @ transformer.ts?d0de:172load @ transformer.ts?d0de:53exports.transform @ transformer.ts?d0de:61(anonymous function) @ transformer.ts?d0de:153transformChildren @ transformer.ts?d0de:151(anonymous function) @ transformer.ts?d0de:172load @ transformer.ts?d0de:53exports.transform @ transformer.ts?d0de:61(anonymous function) @ transformer.ts?d0de:153transformChildren @ transformer.ts?d0de:151(anonymous function) @ transformer.ts?d0de:172load @ transformer.ts?d0de:53exports.transform @ transformer.ts?d0de:61(anonymous function) @ transformer.ts?d0de:153transformChildren @ transformer.ts?d0de:151(anonymous function) @ transformer.ts?d0de:172load @ transformer.ts?d0de:53exports.transform @ transformer.ts?d0de:61(anonymous function) @ transformer.ts?d0de:153transformChildren @ transformer.ts?d0de:151(anonymous function) @ transformer.ts?d0de:172load @ transformer.ts?d0de:53exports.transform @ transformer.ts?d0de:61(anonymous function) @ transformer.ts?d0de:153transformChildren @ transformer.ts?d0de:151(anonymous function) @ transformer.ts?d0de:172load @ transformer.ts?d0de:53exports.transform @ transformer.ts?d0de:61exports.transformToTree @ mutable-tree-factory.ts?f7a1:12(anonymous function) @ mutable-tree-factory.ts?f7a1:40exports.createTreeFromElements @ mutable-tree-factory.ts?f7a1:40updateTree @ backend.ts:64update @ backend.ts:83(anonymous function) @ backend.ts:94SafeSubscriber.__tryOrUnsub @ Subscriber.js?215e:223SafeSubscriber.next @ Subscriber.js?215e:172Subscriber._next @ Subscriber.js?215e:125Subscriber.next @ Subscriber.js?215e:89DebounceTimeSubscriber.debouncedNext @ debounceTime.js?500c:98dispatchNext @ debounceTime.js?500c:114AsyncAction._execute @ AsyncAction.js?e414:111AsyncAction.execute @ AsyncAction.js?e414:86AsyncScheduler.flush @ AsyncScheduler.js?ae69:36ZoneDelegate.invokeTask @ zone.js:365Zone.runTask @ zone.js:265ZoneTask.invoke @ zone.js:432setInterval (async)scheduleTask @ zone.js:1320ZoneDelegate.scheduleTask @ zone.js:345Zone.scheduleMacroTask @ zone.js:282(anonymous function) @ zone.js:1335setInterval @ VM6648:3AsyncAction.requestAsyncId @ AsyncAction.js?e414:65AsyncAction.schedule @ AsyncAction.js?e414:60Scheduler.schedule @ Scheduler.js?bb52:43DebounceTimeSubscriber._next @ debounceTime.js?500c:89Subscriber.next @ Subscriber.js?215e:89Subject.next @ Subject.js?c1c6:55(anonymous function) @ backend.ts:91schedulerFn @ async.ts:162SafeSubscriber.__tryOrUnsub @ Subscriber.ts:240SafeSubscriber.next @ Subscriber.ts:192Subscriber._next @ Subscriber.ts:133Subscriber.next @ Subscriber.ts:93Subject._finalNext @ Subject.ts:154Subject._next @ Subject.ts:144Subject.next @ Subject.ts:90EventEmitter.emit @ async.ts:133(anonymous function) @ ng_zone.ts:158ZoneDelegate.invoke @ zone.js:332Zone.run @ zone.js:225NgZoneImpl.runOuter @ ng_zone_impl.ts:106NgZone.runOutsideAngular @ ng_zone.ts:239NgZone._checkStable @ ng_zone.ts:158onLeave @ ng_zone.ts:135onInvokeTask @ ng_zone_impl.ts:63ZoneDelegate.invokeTask @ zone.js:364Zone.runTask @ zone.js:265ZoneTask.invoke @ zone.js:432XMLHttpRequest.send (async)scheduleTask @ zone.js:101ZoneDelegate.scheduleTask @ zone.js:345Zone.scheduleMacroTask @ zone.js:282(anonymous function) @ zone.js:122send @ VM6659:3(anonymous function) @ xhr_backend.ts:114Observable.subscribe @ Observable.ts:93Observable._subscribe @ Observable.ts:152MapOperator.call @ map.ts:54Observable.subscribe @ Observable.ts:93Observable._subscribe @ Observable.ts:152CatchOperator.call @ catch.ts:32Observable.subscribe @ Observable.ts:93Observable._subscribe @ Observable.ts:152CatchOperator.call @ catch.ts:32Observable.subscribe @ Observable.ts:93ConnectableObservable.connect @ ConnectableObservable.ts:37RefCountObservable._subscribe @ ConnectableObservable.ts:91Observable.subscribe @ Observable.ts:93subscribeToResult @ subscribeToResult.ts:32CombineLatestSubscriber._complete @ combineLatest.ts:184Subscriber.complete @ Subscriber.ts:120ArrayObservable._subscribe @ ArrayObservable.ts:126Observable.subscribe @ Observable.ts:93Observable._subscribe @ Observable.ts:152CombineLatestOperator.call @ combineLatest.ts:150Observable.subscribe @ Observable.ts:93Observable._subscribe @ Observable.ts:152MapOperator.call @ map.ts:54Observable.subscribe @ Observable.ts:93Observable._subscribe @ Observable.ts:152CatchOperator.call @ catch.ts:32Observable.subscribe @ Observable.ts:93subscribeToResult @ subscribeToResult.ts:32SwitchMapSubscriber._innerSub @ switchMap.ts:109SwitchMapSubscriber._next @ switchMap.ts:101Subscriber.next @ Subscriber.ts:93MapSubscriber._next @ map.ts:84Subscriber.next @ Subscriber.ts:93FilterSubscriber._next @ filter.ts:91Subscriber.next @ Subscriber.ts:93Subject._finalNext @ Subject.ts:154Subject._next @ Subject.ts:144ReplaySubject._next @ ReplaySubject.ts:30Subject.next @ Subject.ts:90(anonymous function) @ state-updates.js:30SafeSubscriber.__tryOrUnsub @ Subscriber.ts:240SafeSubscriber.next @ Subscriber.ts:192Subscriber._next @ Subscriber.ts:133Subscriber.next @ Subscriber.ts:93WithLatestFromSubscriber._next @ withLatestFrom.ts:135Subscriber.next @ Subscriber.ts:93Subject._finalNext @ Subject.ts:154Subject._next @ Subject.ts:144SyncSubject._next @ SyncSubject.js:22Subject.next @ Subject.ts:90Dispatcher.dispatch @ dispatcher.js:14Store.dispatch @ store.js:23CommonBootstrapper.bootstrap @ common.bootstrap.ts:74CommonBootstrapper.configure @ common.bootstrap.ts:61AppComponent.bootstrap @ app.component.ts:166AppComponent.ngOnInit @ app.component.ts:115_View_AppComponent_Host0.detectChangesInternal @ AppComponent.template.js:29AppView.detectChanges @ view.ts:261DebugAppView.detectChanges @ view.ts:377AppView.detectContentChildrenChanges @ view.ts:280AppView.detectChangesInternal @ view.ts:272AppView.detectChanges @ view.ts:261DebugAppView.detectChanges @ view.ts:377AppView.detectViewChildrenChanges @ view.ts:288_View_AppRoot_Host0.detectChangesInternal @ AppRoot.template.js:37AppView.detectChanges @ view.ts:261DebugAppView.detectChanges @ view.ts:377ViewRef_.detectChanges @ view_ref.ts:109(anonymous function) @ application_ref.ts:479ApplicationRef_.tick @ application_ref.ts:479(anonymous function) @ application_ref.ts:377ZoneDelegate.invoke @ zone.js:332onInvoke @ ng_zone_impl.ts:72ZoneDelegate.invoke @ zone.js:331Zone.run @ zone.js:225NgZoneImpl.runInner @ ng_zone_impl.ts:104NgZone.run @ ng_zone.ts:219(anonymous function) @ application_ref.ts:377schedulerFn @ async.ts:148SafeSubscriber.__tryOrUnsub @ Subscriber.ts:240SafeSubscriber.next @ Subscriber.ts:192Subscriber._next @ Subscriber.ts:133Subscriber.next @ Subscriber.ts:93Subject._finalNext @ Subject.ts:154Subject._next @ Subject.ts:144Subject.next @ Subject.ts:90EventEmitter.emit @ async.ts:133NgZone._checkStable @ ng_zone.ts:152setMicrotask @ ng_zone.ts:139onHasTask @ ng_zone_impl.ts:85ZoneDelegate.hasTask @ zone.js:391ZoneDelegate._updateTaskCount @ zone.js:408ZoneDelegate.invokeTask @ zone.js:369Zone.runTask @ zone.js:265drainMicroTaskQueue @ zone.js:491ZoneTask.invoke @ zone.js:435XMLHttpRequest.send (async)scheduleTask @ zone.js:101ZoneDelegate.scheduleTask @ zone.js:345Zone.scheduleMacroTask @ zone.js:282(anonymous function) @ zone.js:122send @ VM6659:3XHRImpl.get @ xhr_impl.ts:45DirectiveNormalizer._fetch @ directive_normalizer.ts:51DirectiveNormalizer.normalizeTemplateAsync @ directive_normalizer.ts:98DirectiveNormalizer.normalizeDirective @ directive_normalizer.ts:68RuntimeCompiler._getCompiledTemplate @ runtime_compiler.ts:121RuntimeCompiler._getTransitiveCompiledTemplates @ runtime_compiler.ts:131(anonymous function) @ runtime_compiler.ts:135RuntimeCompiler._getTransitiveCompiledTemplates @ runtime_compiler.ts:134RuntimeCompiler.compileComponentAsync @ runtime_compiler.ts:47RuntimeCompiler.resolveComponent @ runtime_compiler.ts:43AppRoot.ngAfterViewInit @ app.js:272_View_AppRoot_Host0.detectChangesInternal @ AppRoot.template.js:40AppView.detectChanges @ view.ts:261DebugAppView.detectChanges @ view.ts:377ViewRef_.detectChanges @ view_ref.ts:109(anonymous function) @ application_ref.ts:479ApplicationRef_.tick @ application_ref.ts:479ApplicationRef_._loadComponent @ application_ref.ts:453(anonymous function) @ application_ref.ts:439(anonymous function) @ application_ref.ts:407ZoneDelegate.invoke @ zone.js:332onInvoke @ ng_zone_impl.ts:72ZoneDelegate.invoke @ zone.js:331Zone.run @ zone.js:225NgZoneImpl.runInner @ ng_zone_impl.ts:104NgZone.run @ ng_zone.ts:219ApplicationRef_.run @ application_ref.ts:405ApplicationRef_.bootstrap @ application_ref.ts:429(anonymous function) @ application_ref.ts:169ZoneDelegate.invoke @ zone.js:332onInvoke @ ng_zone_impl.ts:72ZoneDelegate.invoke @ zone.js:331Zone.run @ zone.js:225(anonymous function) @ zone.js:586ZoneDelegate.invokeTask @ zone.js:365onInvokeTask @ ng_zone_impl.ts:61ZoneDelegate.invokeTask @ zone.js:364Zone.runTask @ zone.js:265drainMicroTaskQueue @ zone.js:491Promise.resolve (async)scheduleQueueDrain @ zone.js:464scheduleMicroTask @ zone.js:472ZoneDelegate.scheduleTask @ zone.js:348Zone.scheduleMicroTask @ zone.js:279scheduleResolveOrReject @ zone.js:584ZoneAwarePromise.then @ zone.js:662ZoneAwarePromise.all @ zone.js:641PromiseWrapper.all @ promise.ts:37(anonymous function) @ application_ref.ts:365(anonymous function) @ application_ref.ts:407ZoneDelegate.invoke @ zone.js:332onInvoke @ ng_zone_impl.ts:72ZoneDelegate.invoke @ zone.js:331Zone.run @ zone.js:225NgZoneImpl.runInner @ ng_zone_impl.ts:104NgZone.run @ ng_zone.ts:219ApplicationRef_.run @ application_ref.ts:405ApplicationRef_ @ application_ref.ts:351(anonymous function) @ reflection_capabilities.ts:31ReflectiveInjector_._instantiate @ reflective_injector.ts:755ReflectiveInjector_._instantiateProvider @ reflective_injector.ts:683ReflectiveInjector_._new @ reflective_injector.ts:672ReflectiveInjectorDynamicStrategy.getObjByKeyId @ reflective_injector.ts:305ReflectiveInjector_._getByKeyDefault @ reflective_injector.ts:871ReflectiveInjector_._getByKey @ reflective_injector.ts:840ReflectiveInjector_._getByReflectiveDependency @ reflective_injector.ts:824ReflectiveInjector_._instantiate @ reflective_injector.ts:715ReflectiveInjector_._instantiateProvider @ reflective_injector.ts:683ReflectiveInjector_._new @ reflective_injector.ts:672ReflectiveInjectorDynamicStrategy.getObjByKeyId @ reflective_injector.ts:305ReflectiveInjector_._getByKeyDefault @ reflective_injector.ts:871ReflectiveInjector_._getByKey @ reflective_injector.ts:840ReflectiveInjector_.get @ reflective_injector.ts:633coreLoadAndBootstrap @ application_ref.ts:164bootstrap @ index.ts:126(anonymous function) @ bootstrap.js:38cssReady @ bootstrap.js:63(anonymous function) @ bootstrap.js:68
```

**Comment by tb** [comment]
I have same error on Angular2 2.0.0-rc.5

    Uncaught TypeError: Cannot read property 'name' of undefined
    exports.functionName	@	function-name.ts?6954:7

**Pull request merged**